src/expr/term_context.h

```diff
@@ -0,0 +1,143 @@
+ *   (and P (not Q))
+ * are P @ [0] and Q @ [1,0].
+ */
+class TermContext
+{
+ public:
+  TermContext() {}
+  virtual ~TermContext() {}
+  /** The default initial value of root terms. */
+  virtual uint32_t initialValue() const = 0;
+  /**
+   * Returns the term context identifier of the index^th child of t, where tval
+   * is the term context identifier of t.
+   */
+  virtual uint32_t computeValue(TNode t, uint32_t tval, size_t index) const = 0;
+};
+
+/**
+ * Remove term formulas (rtf) term context.
+ *
+ * Computes whether we are inside a term (as opposed to being part of Boolean
+ * skeleton) and whether we are inside a quantifier. For example, for:
+ *   (and (= a b) (forall ((x Int)) (P x)))
+ * we have the following mappings (term -> inTerm,inQuant)
+ *   (= a b) @ [0] -> false, false
+ *   a @ [0,1] -> true, false
+ *   (P x) @ [1,1] -> false, true
+ *    x @ [1,1,0] -> true, true
+ * Notice that the hash of a child can be computed from the parent's hash only,
+ * and hence this can be implemented without considering paths.
+ */
+class RtfTermContext : public TermContext
+{
+ public:
+  RtfTermContext() {}
+  /** The initial value: not in a term context or beneath a quantifier. */
+  uint32_t initialValue() const override;
+  /** Compute the value */
+  uint32_t computeValue(TNode t, uint32_t tval, size_t index) const override;
+  /** get hash value from the flags */
+  static uint32_t getValue(bool inQuant, bool inTerm);
+  /** get flags from the hash value */
+  static void getFlags(uint32_t val, bool& inQuant, bool& inTerm);
+
+ private:
+  /**
+   * Returns true if the children of t should be considered in a "term" context.
+   */
+  static bool hasNestedTermChildren(TNode t);
+};
+
+/**
+ * Polarity term context.
+ *
+ * This class computes the polarity of a term-context-sensitive term, which is
+ * one of {true, false, none}. This corresponds to the value that can be
+ * assigned to that term while preservering satisfiability of the overall
+ * formula, or none if such a value does not exist. If not "none", this
+ * typically corresponds to whether the number of NOT the formula is beneath is
+ * even, although special cases exist (e.g. the first child of IMPLIES).
+ *
+ * For example, given the formula:
+ *   (and P (not (= (f x) 0)))
+ * assuming the root of this formula has true polarity, we have that:
+ *   P @ [0] -> true
+ *   (not (= (f x) 0)) @ [1] -> true
+ *   (= (f x) 0) @ [1,0] -> false
+ *   (f x) @ [1,0,0]), x @ [1,0,0,0]), 0 @ [1,0,1] -> none
+ *
+ * Notice that a term-context-sensitive Node is not one-to-one with Node.
+ * In particular, given the formula:
+ *   (and P (not P))
+ * We have that the P at path [0] has polarity true and the P at path [1,0] has
+ * polarity false.
+ *
+ * Finally, notice that polarity does not correspond to an entailed value. Thus,
+ * for the formula:
+ *   (or P Q)
+ * we have that
+ *   P @ [0] -> true
+ *   Q @ [1] -> true
+ * although neither is entailed.
```

Comment:
Not sure what this is supposed to convey. What would be an "entailed value"?

Reply:
A value that the formula entails.

Reply:
I see. I think this more confuses than clarifies things if one does not expect polarity to correspond to entailed values.

Reply:
This is an important distinction, there will be another variant of this class called EntailedPolarityTermContext. This comment is in preparation for that class, where this distinction will be important.

Reply:
I see.